### PR TITLE
travis osx support

### PR DESCRIPTION
**Pull request by wxchan**
#578

### CODE REVIEW

**Comment by guolinke** [comment]
@wxchan The gpu build failed.
BTW, can we add a task named "pip", and test the build of source package ?

**Comment by wxchan** [comment]
@guolinke do you know what is lacking for gpu task? I copied old configuration to here.
The error is:
```
/home/travis/build/Microsoft/LightGBM/compute/include/boost/compute/cl.hpp:17:19: fatal error: CL/cl.h: No such file or directory
 #include <CL/cl.h>
```

**Comment by wxchan** [comment]
for TASK=pip, the error actually happens every time before, but I don't know why it passes, you can see the build from master branch, Line 1837.
https://travis-ci.org/Microsoft/LightGBM/jobs/245330471

**Comment by guolinke** [comment]
@wxchan fixed in the latest master. Can you merge it ? 
BTW, why didn't enable pip test in osx ?

**Comment by guolinke** [comment]
@wxchan 
for the gpu build, it seems the opencl cannot be found.

master log: https://travis-ci.org/Microsoft/LightGBM/builds/245646792#L1177

**Comment by wxchan** [comment]
Why travis sometimes is not enabled?

File: .travis/test.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    pep8 --ignore=E501 --exclude=./compute,./docs . || exit -1
+    exit 0
+fi
+
+conda install --yes numpy scipy scikit-learn pandas matplotlib
+pip install pytest
+
+if [[ $TRAVIS_OS_NAME == "linux" ]]; then 
+    conda install --yes atlas
+    conda install --yes -c conda-forge boost=1.63.0
```
Comment (guolinke) [inline .travis/test.sh]: I think the boost is not needed for the non-gpu tests.

Comment (guolinke) [inline .travis/test.sh]: not sure about atlas, which is used in numpy ?

Comment (wxchan) [inline .travis/test.sh]: yes, not sure it's required, I try to remove it, see if it works

**Comment by wxchan** [comment]
we can refer to this doc for pypi release: https://docs.travis-ci.com/user/deployment/pypi/

**Comment by guolinke** [comment]
@wxchan yeah. And we can create a branch for the deploy, or the the build will be too frequently .

**Comment by wxchan** [comment]
Test passes. It saves about half of time on other task.

For osx, it spends a lot of time running `brew install gcc`, it will be good to have an alternative way.

**Comment by guolinke** [comment]
@wxchan 
Now the default cpp compiler in OSX still cannot support openmp ?

**Comment by guolinke** [comment]
@wxchan 
BTW, the test of if...else conversion is removed ?

**Comment by wxchan** [comment]
@guolinke on osx, it even can't find gcc. I think it has something to do with gcc version.

I think use `brew install gcc` is fine for now, ~8 mins is not too long. Let's leave it for future refine (maybe when add R tests).

**Comment by wxchan** [comment]
`.travis/test.sh: line 54: ../../lightgbm: No such file or directory`

where is lightgbm binary file now after `make _lightgbm`?

**Comment by guolinke** [comment]
@wxchan 
use ```make lightgbm``` to build binary. And I think the numpy sklean, ... , is not needed is if ... else test.

Actually, I doubt the multi threading performance of gcc in mac osx. Maybe clang with openmp is better.

**Comment by wxchan** [comment]
@guolinke 
https://github.com/wxchan/LightGBM/pull/1/files

I tried using clang with openmp for mac, but it could not find OpenMP. On my mac, there are two versions of clang-omp, when I guess it's installed by `brew install clang-omp` long time ago, work perfectly but deprecated. This one installed by `brew install llvm` got errors.

**Comment by guolinke** [comment]
@wxchan can you test the performance between clang-omp and gcc ?

**Pull request merged**